src/abi/substrate.rs

```diff
@@ -1,284 +1,459 @@
+                let path = registry.get(ty).unwrap().path().clone();
+                Some(TypeSpec::new(ty.into(), path))
             }
-        }
-    }
-
-    /// Returns index to builtin type in registry. Type is added if not already present
-    fn builtin_type(&mut self, ty: &str) -> usize {
-        self.register_ty(Type::Builtin {
-            def: PrimitiveDef {
-                primitive: ty.to_owned(),
-            },
-        })
-    }
+            _ => {
+                let fields = f
+                    .returns
+                    .iter()
+                    .map(|r_p| {
+                        let ty = resolve_ast(&r_p.ty, ns, &mut registry);
+
+                        ty.into()
+                    })
+                    .collect::<Vec<_>>();
+
+                let t = TypeDefTuple::new_portable(fields);
+
+                let path = path!(
+                    &ns.contracts[contract_no].name,
+                    &f.name,
+                    &"return_type".into()
+                );
+
+                let ty = registry.register_type(Type::new(
+                    path,
+                    vec![],
+                    TypeDef::Tuple(t),
+                    Default::default(),
+                ));
+                let path = registry.get(ty).unwrap().path().clone();
+                Some(TypeSpec::new(ty.into(), path))
+            }
+        };
+        let ret_type = ReturnTypeSpec::new(ret_spec);
+        let args = f
+            .params
+            .iter()
+            .map(|p| {
+                let ty = resolve_ast(&p.ty, ns, &mut registry);
+                let path = registry.get(ty).unwrap().path().clone();
+                let spec = TypeSpec::new(ty.into(), path);
+
+                MessageParamSpec::new(p.name_as_str().to_string())
+                    .of_type(spec)
+                    .done()
+            })
+            .collect::<Vec<MessageParamSpec<PortableForm>>>();
+        let label = if conflicting_names.contains(&f.name) {
+            &f.mangled_name
+        } else {
+            &f.name
+        };
+        MessageSpec::from_label(label.into())
+            .selector(f.selector().try_into().unwrap())
+            .mutates(mutates)
+            .payable(payable)
+            .args(args)
+            .returns(ret_type)
+            .docs(vec![render(&f.tags)])
+            .done()
+    };
 
-    /// Returns index to builtin type in registry. Type is added if not already present
-    fn builtin_array_type(&mut self, elem: usize, array_len: usize) -> usize {
-        self.register_ty(Type::BuiltinArray {
-            def: ArrayDef {
-                array: Array {
-                    len: array_len,
-                    ty: elem,
-                },
-            },
+    let messages = ns.contracts[contract_no]
+        .all_functions
+        .keys()
+        .filter_map(|function_no| {
+            let func = &ns.functions[*function_no];
+            // escape if it's a library
```

Comment:
maybe this should say something like `// libraries are never in the public interface`